Source/JavaScriptCore/wasm/js/JSWebAssembly.cpp

```diff
@@ -224,7 +223,14 @@ static void instantiate(VM& vm, JSGlobalObject* globalObject, JSPromise* promise
                 break;
             }
             }
-        });
+        };
+
+        if (alwaysAsync || isAsync) {
+            vm.deferredWorkTimer->scheduleWorkSoon(ticket, WTFMove(callback));
+            return;
+        }
+        vm.deferredWorkTimer->cancelPendingWork(ticket);
+        callback(ticket);
```

Comment:
I am confusing here. Also, for the previous implementation. My question here is that at this point, we know that it's compiled. So, why can't we just do 
```
callback(ticket);
```
rather than using `deferredWorkTimer`. If the purpose of using `deferredWorkTimer` is keep the corresponding `dependencies` alive. Then, why is it OK to do `callback(ticket);` here? 

Does that mean when `!alwaysAsync && !isAsync` then it must be in a sync path. And at that point we guarantee all the `dependencies` are alive. Thus, it's OK to cancel the tick and do the callback immediately.

Reply:
The reason is that compilation callback can be called at non main thread. So we need to schedule a task to post a task to the main thread. But sometimes, it is not necessary and that's how these boolean flags are managing.